### PR TITLE
Sort changelogs by date then version

### LINKED ISSUE #1389: Release script should sort by date first, version second
Sometimes we need to release hotfixes, so older versions may ship after newer ones like in Key Vault.

```diff
 - Default service version is now 7.1.

+## 4.1.0-preview.1 (2020-03-09)
+
+### Added
+
+- Add `RecoverableDays` property to `CertificateProperties`.
+
 ## 4.0.3 (2020-07-09)

 ### Fixed
@@ -56,12 +62,6 @@

 - Fixed concurrency issue in our challenge-based authentication policy ([#9737](https://github.com/Azure/azure-sdk-for-net/issues/9737))

-## 4.1.0-preview.1 (2020-03-09)
-
-### Added
-
-- Add `RecoverableDays` property to `CertificateProperties`.
-
 ## 4.0.1 (2020-03-03)
```

The _Prepare-Release.ps1_ release script should appropriately sort sections by date first, then release version. CHANGELOGs typically happen in chronological order.

### PR DESCRIPTION
**Pull request by heaths**
Fixes #1389

### CODE REVIEW

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 1]
```diff
@@ -193,15 +193,14 @@ function Set-ChangeLogContent {
 
   try
   {
-    $VersionsSorted = [AzureEngSemanticVersion]::SortVersionStrings($ChangeLogEntries.Keys)
+    $ChangeLogEntries = $ChangeLogEntries.Values | Sort-Object -Descending -Property ReleaseStatus, ReleaseVersion
```
Comment (heaths) [inline eng/common/scripts/ChangeLog-Operations.ps1:196]: By sorting by `ReleaseStatus` (date or "(Unreleased)") first, it actually means any older entries without dates get sorted earlier in the final content. Probably worth a one-time fixup wherever necessary for customers' benefit.

Comment (weshaggard) [reply eng/common/scripts/ChangeLog-Operations.ps1:196]: Yes I'm fine with having some older entries getting manually fixed up, we were really already depending on that.

With this change does "(Unreleased)" end up sorting to the top?

File: eng/common/scripts/SemVer.ps1 [hunk 2]
```diff
@@ -187,10 +212,7 @@ class AzureEngSemanticVersion {
...
   static [AzureEngSemanticVersion[]] SortVersions([AzureEngSemanticVersion[]] $versions)
   {
-    return ($versions | `
-            Sort-Object -Descending -Property `
-              Major, Minor, Patch, PrereleaseLabel, PrereleaseNumber, `
-              @{ Expression = { [int]$_.BuildNumber }; Descending = $true })
+    return $versions | Sort-Object -Descending
   }
 
   static [void] QuickTests()
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:218]: If you haven't already could you please do a quick run of this to make sure the basic tests we have still work the same?

**Review by weshaggard — APPROVED** [review]
Thanks for fixing this @heaths.

File: eng/common/scripts/SemVer.ps1 (hunk 2 above)
Comment (heaths) [reply eng/common/scripts/SemVer.ps1:218]: I did, yes. In fact, I had to to debug an issue I was having in the original iteration.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 1 above)
Comment (heaths) [reply eng/common/scripts/ChangeLog-Operations.ps1:196]: Yep. Fortunately letters sort after numbers, so when we reverse the order it works out nicely. :)

File: eng/common/scripts/SemVer.ps1 [hunk 3]
```diff
@@ -13,7 +13,7 @@ Components: Major.Minor.Patch-PrereleaseLabel.PrereleaseNumber.BuildNumber
 Note: A builtin Powershell version of SemVer exists in 'System.Management.Automation'. At this time, it does not parsing of PrereleaseNumber. It's name is also type accelerated to 'SemVer'.
 #>
 
-class AzureEngSemanticVersion {
+class AzureEngSemanticVersion : IComparable {
```
[the remark this answers is not in the document]
Comment (danieljurek) [reply eng/common/scripts/SemVer.ps1:16]: This is a much better (more readable) version than the one I wrote. I vote we use this one. Did some local sanity checking and this implementation checks out.

[the remark this answers is not in the document]
Comment (heaths) [reply eng/common/scripts/SemVer.ps1:16]: Initially I tried implementing the generic `IComparable[AzureEngSemanticVersion]` interface, but seems you can't do that in PowerShell given the order of when types are resolved vs. compiled.

**Comment by heaths** [comment]
@weshaggard @chidozieononiwu @danieljurek should I send out an email about this, and make sure people know to add (preferably accurate) dates for older versions in their changelogs?

**Comment by weshaggard** [comment]
I don't think it is worth sending an email as most folks will ignore it. If there are any issues they will come to light in the version increment PRs.

**Pull request merged**